### ISSUE TITLE
Not working if there's ipython magic function in cell

### ISSUE THREAD

**Issue by Globefishp** [EeyoreLee/vscode-extension-yapf#39]
In Jupyter Notebook(.ipynb) file, When a cell contains magic function such as `%matplotlib inline`, yapf cannot format this cell, but autopep8(builtin formatter) can.
When I delete `%matplotlib inline`, yapf works fast and well.
I hope this issue will be fix soon.
Thanks

**Comment by Globefishp** [comment]
VS Code version: 1.82.2, 
Jupyter Extension Version: v2023.8.1002501831

**Comment by EeyoreLee** [comment]
@Globefishp - hi, I've known about this problem cause google/yapf did't support for jupyter magic line, related issue [google/yapf/issue#218](https://github.com/google/yapf/issues/218). In fact, it should be the google/yapf's responsibility. Well, I plan to do this feature like black package through adding a property that users can add a magic function list to avoid breaking the whole cell. I saw some other formatter extension which support this choose to just replace `%` to another like `#%#`, and then recovery, but it will make some python code line get a ugly formating like 
```
1\
%2
```

**Comment by EeyoreLee** [comment]
@Globefishp - I released a new version v0.1.9 which supported jupyter magic function with some default below.
```
"capture",
"prun",
"pypy",
"python",
"python3",
"time",
"timeit"
```
other magic commands like `%matplotlib inline`, you need to add it to the `yapf.cellMagics` property.
```
"yapf.cellMagics": ["matplotlib inline"]
```

**Comment by Globefishp** [comment]
Thanks for your quick reply! I think just ignore those magic function and don’t pass them to yapf is a good idea. To improve performance, it will be good to only activate these rules if formatting .ipynb files.

I also wonder if your yapf extension can implement “Format on Type” feature? As far as I know, now not a single python formatter in VS Code can do that. About 2 years ago there’s official support built in python extension of VS Code (but now it’s broken, there’s no difference whether you switch on/off this in setting.) It’s quite simple as it just format current line as you press Enter. (but I don’t know if there’s other obstacle to add this feature.)

Thanks!

________________________________
发件人: Earlee ***@***.***>
发送时间: Friday, September 22, 2023 11:56:38 AM
收件人: EeyoreLee/vscode-extension-yapf ***@***.***>
抄送: Globefishp ***@***.***>; Mention ***@***.***>
主题: Re: [EeyoreLee/vscode-extension-yapf] Not working if there's ipython magic function in cell (Issue #39)


@Globefishp<https://github.com/Globefishp> - hi, I've known about this problem cause google/yapf did't support for jupyter magic line, related issue google/yapf/issue#218<https://github.com/google/yapf/issues/218>. In fact, it should be the google/yapf's responsibility. Well, I plan to do this feature like black package through adding a property that users can add a magic function list to avoid breaking the whole cell. I saw some other formatter extension which support this choose to just replace % to another like #%#, and then recovery, but it will make some python code line get a ugly formating like

1\
%2


―

**Comment by Globefishp** [comment]
Thanks for your efforts!

________________________________
发件人: Earlee ***@***.***>
发送时间: Monday, September 25, 2023 6:15:43 PM
收件人: EeyoreLee/vscode-extension-yapf ***@***.***>
抄送: Globefishp ***@***.***>; Mention ***@***.***>
主题: Re: [EeyoreLee/vscode-extension-yapf] Not working if there's ipython magic function in cell (Issue #39)


@Globefishp<https://github.com/Globefishp> - I released a new version v0.1.9 which supported jupyter magic function with some default below.

"capture",
"prun",
"pypy",
"python",
"python3",
"time",
"timeit"


other magic commands like %matplotlib inline, you need to add it to the yapf.cellMagics property.

"yapf.cellMagics": ["matplotlib inline"]


―

**Comment by EeyoreLee** [comment]
@Globefishp - hi, I wonder if you edit the code below on formatOnType mode
```
x = 1 + \
2+3
```
What results do you expect? If when you edited the second line, and press enter, `2+3` will format to `2 + 3`, which is different from format the snippet. Even if you put a backslash at the tuple or list like 
```
x = (1, 2, \
3)
```
the only second line `3)` will raise syntax error. The reason why I didn't implement it is just I didn't have an idea that what result is the expected indeed.

**Comment by Globefishp** [comment]
I think I’ve got your idea. Though these code is definitely right, I have never coded like that. I agree with you that there’s some kinds of ambiguity to format those code. I think mostly in this case, it is the user that wishes to control the format and alignment. So it’s good not to format these kinds of code (explicitly split one line to multiple line by “\”) by default (in formatOnType mode).

Another case is a code like:
```
    dW2 = torch.zeros((*X.shape, num_classes),
                      dtype=torch.float64,
                      device='cuda')
```
I think in this case, the formatter should detect if there’s a syntax error. If there’s no error, the code will be merged as a single line, formatted and further splitted by “group of arguments”. That means:
```
    dW2=torch.zeros((*X.shape,num_classes),dtype=torch.float64,
                            device='cuda')
```
will result:
```
    dW2 = torch.zeros((*X.shape, num_classes), dtype=torch.float64,
                      device='cuda')
```
And
```
    dW2 = torch.zeros((*X.shape, num_classes),
                                     dtype=torch.float64,
                             device='cuda')
```
will result:
```
    dW2 = torch.zeros((*X.shape, num_classes),
                      dtype=torch.float64,
                      device='cuda')
```

During typing, if I press ENTER after `dtype=torch.float64, `, there is syntax error so the formatter do nothing. Once I fininsh this line and press ENTER after `device='cuda')`, this line is completed, and it will get formatted.

That's basically two cases I could think out. I don't know if there's more cases should take into consideration.

**Comment by EeyoreLee** [comment]
I think the most basic principle is - getting the same results in different modes. I need to take a look at the new lsp, and the other extensions like black which follow lsp also didn't support format on type, even didn't support on modifications mode. I will add this feature on next release.

**Comment by Globefishp** [comment]
It will be good if there's a way to make those results consistent in different modes. Thanks again for considering this feature.

**Comment by EeyoreLee** [comment]
@Globefishp - hi, the version v0.1.10 release with `FormatOnType` feature.

**Comment by Globefishp** [comment]
Hi, I've tried v0.1.10 and find it support some predefined (such as prun) magic function really well, but broken even if I directly added a setting `"yapf.cellMagics": ["matplotlib inline"]` or add through GUI provided in vscode (which add a setting string like
```
    "yapf.cellMagics": [
        "[\"matplotlib inline\"]"
    ]
```
My code is in .ipynb file and looks like:
```
%matplotlib inline
import numpy as np
steps=1000
```
if I comment `%matplotlib inline`, everything works well.
I don't know what's going on. I've tried to turn off `formatOnType`, but that doesn't help.
I wonder if I have a wrong configuration somewhere?

Thanks.

**Comment by EeyoreLee** [comment]
@Globefishp - There's no need to nest a list. 
```
"yapf.cellMagics": ["matplotlib inline"]
```
The value of property is just like python list, and the elem of list is the magic function name.

**Comment by EeyoreLee** [comment]
It seems python extension won't restart when turn off `formatOnType`, so sometimes you can just do a harmless change for any yapf property to let the `yapf` extension restart. After restart message output on the yapf channel, `formatOnType` is really turn off.But it's careless with `formatOnSave`, I think the "yapf.cellMagics" property is not be setted correctly as I just test it on my side and it works fine.

**Comment by Globefishp** [comment]
Here's my output. I wonder if this is a correct setting.
```
2023-10-27 10:54:54.139 [info] Name: yapf
2023-10-27 10:54:54.139 [info] Module: yapf
2023-10-27 10:54:54.139 [info] Python extension loading
2023-10-27 10:54:54.139 [info] Waiting for interpreter from python extension.
2023-10-27 10:54:54.139 [info] Python extension loaded
2023-10-27 10:54:54.139 [info] Server run command: c:\Users\Globefish\Miniconda3\python.exe c:\Users\Globefish\.vscode\extensions\eeyore.yapf-0.1.10\bundled\tool\lsp_server.py
2023-10-27 10:54:54.139 [info] Server: Start requested.
2023-10-27 10:54:54.987 [info] CWD Server: d:\GraduateSchool\2023\数理基础\HOMEWORK-4\Code
2023-10-27 10:54:54.987 [info] sys.path used to run Server:
   c:\Users\Globefish\.vscode\extensions\eeyore.yapf-0.1.10\bundled\libs
   c:\Users\Globefish\.vscode\extensions\eeyore.yapf-0.1.10\bundled\tool
   c:\Users\Globefish\Miniconda3\python39.zip
   c:\Users\Globefish\Miniconda3\DLLs
   c:\Users\Globefish\Miniconda3\lib
   c:\Users\Globefish\Miniconda3
   c:\Users\Globefish\Miniconda3\lib\site-packages
   c:\Users\Globefish\Miniconda3\lib\site-packages\win32
   c:\Users\Globefish\Miniconda3\lib\site-packages\win32\lib
   c:\Users\Globefish\Miniconda3\lib\site-packages\Pythonwin
   c:\Users\Globefish\.vscode\extensions\eeyore.yapf-0.1.10\bundled\tool-libs
2023-10-27 10:54:54.987 [info] Settings used to run Server:
[
    {
        "cwd": "d:\\GraduateSchool\\2023\\数理基础\\HOMEWORK-4\\Code",
        "workspace": "file:///d%3A/GraduateSchool/2023/%E6%95%B0%E7%90%86%E5%9F%BA%E7%A1%80/HOMEWORK-4/Code",
        "args": [],
        "path": [],
        "interpreter": [
            "c:\\Users\\Globefish\\Miniconda3\\python.exe"
        ],
        "importStrategy": "fromEnvironment",
        "showNotifications": "off",
        "showDebugLog": false,
        "cellMagics": [
            "matplotlib inline"
        ]
    }
]

2023-10-27 10:54:54.988 [info] Global settings:
{
    "cwd": "C:\\Program Files\\Microsoft VS Code",
    "workspace": "C:\\Program Files\\Microsoft VS Code",
    "args": [],
    "path": [],
    "interpreter": [],
    "importStrategy": "fromEnvironment",
    "showNotifications": "off",
    "showDebugLog": false,
    "cellMagics": [
        "matplotlib inline"
    ]
}

2023-10-27 10:55:15.434 [info] c:\Users\Globefish\.vscode\extensions\eeyore.yapf-0.1.10\bundled\tool\lsp_server.py:88: DeprecationWarning: 'workspace.get_document' has been deprecated, use 'workspace.get_text_document' instead
  document = LSP_SERVER.workspace.get_document(params.text_document.uri)

2023-10-27 10:55:15.434 [info] c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
2023-10-27 10:55:15.435 [info] CWD Linter: .
2023-10-27 10:55:15.513 [info] yapf: <stdin>:1:1: invalid syntax

2023-10-27 10:55:47.556 [info] c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
2023-10-27 10:55:47.556 [info] CWD Linter: .
```
By the way, a nested list is generated by vscode, so I think in the tips, `e.g. ["matplotlib inline"]` should changed to `e.g. matplotlib inline`?
![image](https://github.com/EeyoreLee/vscode-extension-yapf/assets/31174776/830adacb-6ab3-4303-bebd-3f0f1c0b9b9e)

**Comment by EeyoreLee** [comment]
@Globefishp - The setting is correct and the hint message in GUI is wrong which should be fix. Could you set property `yapf.showDebugLog` to `true` and show the output to check whether the uri is start with `vscode-notebook-cell`. Did you press ctrl+s tigger formatting ? And I think it is most likely caused by the python extension as it used `rangeFormat` by `-l` which is shouldn't be.
```
c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
```
so could try to upgrade `Python` extension which `yapf` depend on.

**Comment by EeyoreLee** [comment]
> Hi, I've tried v0.1.10 and find it support some predefined (such as prun) magic function really well, but broken even if I directly added a setting `"yapf.cellMagics": ["matplotlib inline"]` or add through GUI provided in vscode (which add a setting string like
> 
> ```
>     "yapf.cellMagics": [
>         "[\"matplotlib inline\"]"
>     ]
> ```
> 
> My code is in .ipynb file and looks like:
> 
> ```
> %matplotlib inline
> import numpy as np
> steps=1000
> ```
> 
> if I comment `%matplotlib inline`, everything works well. I don't know what's going on. I've tried to turn off `formatOnType`, but that doesn't help. I wonder if I have a wrong configuration somewhere?
> 
> Thanks.

Does it mean it works for other built-in magic function like `%time` for you?

**Comment by EeyoreLee** [comment]
On Windows OS, the following output should appear when `yapf.showDebugLog` is setted to `true`
```
2023-10-27 14:42:54.688 [info] vscode-notebook-cell:/c%3A/Users/sh2993/Desktop/yapf-windows/test.ipynb#W1sZmlsZQ%3D%3D :
%matplotlib inline
import numpy as np

steps = 1000
```

**Comment by Globefishp** [comment]
> yapf.showDebugLog

Yes, %time works well. 
I've upgrade to python extension v2023.20.0, pylance v2023.11.10.
I've tried to open debugOutput, it shows:
```
2023-11-05 02:07:23.068 [info] c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
2023-11-05 02:07:23.069 [info] CWD Linter: .
2023-11-05 02:07:23.073 [info] vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D :
%time
import numpy as np

step = 1000
```
after formatting.

but with `%matplotlib inline`, 
```
2023-11-05 02:09:19.443 [info] c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
2023-11-05 02:09:19.443 [info] CWD Linter: .
2023-11-05 02:09:19.446 [info] yapf: <stdin>:1:1: invalid syntax

2023-11-05 02:09:19.446 [info] vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D :
```
if I comment  `%matplotlib inline`,
```
2023-11-05 02:10:28.989 [info] c:\Users\Globefish\Miniconda3\python.exe -m yapf -l 3-3
2023-11-05 02:10:28.989 [info] CWD Linter: .
2023-11-05 02:10:28.994 [info] vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D :
# %matplotlib inline
import numpy as np

step = 1000
```

so does it have anything to do with my conda and python version? my python is 3.9.18 and conda 23.9.0 which is just up to date.
Thanks again for your help.

**Comment by EeyoreLee** [comment]
@Globefishp - I just published a pre-release version v0.1.12 with a little fix and a debug log for cellMagics. Could you switch it to pre-release version by following step, and show the new output if it still not works for you
![image](https://github.com/EeyoreLee/vscode-extension-yapf/assets/49790022/fffcbefa-1c7e-4bf7-a556-6e7ae348b0bd)
And the next offical release version will be follow `major.minor.patch` format `v0.2.0`

**Comment by Globefishp** [comment]
Thanks, the version 0.1.13 is now working well on my PC ^_^
